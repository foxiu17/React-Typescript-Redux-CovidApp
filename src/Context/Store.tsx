import React, { createContext, useReducer } from "react";

import { IInitialState, IActionType } from "../Interfaces/StoreInterfaces";

const initialState: IInitialState = {
  allData: [],
  countryData: [],
  countries: []
};

export const Store = createContext<IInitialState | any>(initialState);

function Reducers(state: IInitialState, action: IActionType): IInitialState {
  switch (action.type) {
    case "FETCH_COVID_SUMMARY":
      return {
        ...state,
        allData: action.payload
      };
    case "FETCH_COVID_BY_COUNTRY":
      return {
        ...state,
        countryData: action.payload
      };
    case "FETCH_COVID_COUNTRIES":
      return {
        ...state,
        countries: action.payload
      };
    default:
      return state;
  }
}
export function StoreProvider(props: any): JSX.Element {
  const [state, dispatch] = useReducer(Reducers, initialState);
  return <Store.Provider value={{ state, dispatch }}>{props.children}</Store.Provider>;
};