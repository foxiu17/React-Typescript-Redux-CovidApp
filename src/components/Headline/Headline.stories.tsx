import React from "react";

import Headline from "./Headline";

export default {
  component: Headline,
  title: "Headline"
};

export const withoutProps = () => <Headline />;

export const withProps = () => <Headline totalCount={1000} />;