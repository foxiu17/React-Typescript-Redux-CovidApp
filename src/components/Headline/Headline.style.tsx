import styled from 'styled-components';

import { BlockContainer as commonBlockContainer, Grid as commonGrid } from "../../Assets/Common/Containers.style";
import { Title as commonTitle, SubTitle as commonSubTitle } from '../../Assets/Common/Text';

export const Container = styled(commonBlockContainer)`
  padding: 10px 0 50px;
`;

export const Grid = styled(commonGrid)``;

export const Title = styled(commonTitle)`
  font-size: 72px;
  text-align: center;
`;

export const Subtitle = styled(commonSubTitle)`
  font-size: 24px;
  text-align: center;
`;