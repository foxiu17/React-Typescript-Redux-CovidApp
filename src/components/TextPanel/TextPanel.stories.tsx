import React from 'react';

import TextPanel from "./TextPanel";

export default {
  component: TextPanel,
  title: "TextPanel"
};

export const withoutProps = () => <TextPanel />;

export const greenWithProps = () => <TextPanel color="#2ae2ce" title="Total death" count={1000} />;

export const blueWithProps = () => <TextPanel color="#2232ce" title="Total death" count={1000} />;

export const withNotAllProps = () => <TextPanel title="Test title Storybook" />;