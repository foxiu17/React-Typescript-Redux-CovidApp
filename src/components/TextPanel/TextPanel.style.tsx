import styled from 'styled-components';

import { IGridProps } from "../../Interfaces/Styles/StylesInterfaces";

import { Grid as commonGrid } from '../../Assets/Common/Containers.style';

export const Grid = styled(commonGrid)<IGridProps>`
  width: 100%;
  height: 100%;
`;