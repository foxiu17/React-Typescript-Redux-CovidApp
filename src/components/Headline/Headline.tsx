import React from "react";

import { Container, Grid, Title, Subtitle } from "./Headline.style";

interface IHeadlineProps {
  totalCount?: number
}

const Headline = ({ totalCount }: IHeadlineProps) => {
  return (
    <Container>
      <Grid container justify="space-between" spacing={3}>
        <Grid item xs={12}>
          <Title>CORONAVIRUS</Title>
        </Grid>
        <Grid item xs={12}>
          <Subtitle>{totalCount} infected people on all the world!</Subtitle>
        </Grid>
      </Grid>
    </Container>
  )
};

export default Headline;