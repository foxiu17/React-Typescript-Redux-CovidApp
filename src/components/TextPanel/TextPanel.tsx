import React from "react";

import { Grid } from './TextPanel.style';

import { ITextPanelProps } from "../../Interfaces/Interfaces";

const TextPanel = ({ color, title, count }: ITextPanelProps) => {
  return (
    <Grid container justify="center" color={color}>
      <Grid item xs={12}>{title}</Grid>
      <Grid item xs={12}>{count}</Grid>
    </Grid>
  )
};

export default TextPanel;