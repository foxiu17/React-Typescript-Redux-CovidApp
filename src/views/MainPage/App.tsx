import React, { useState, useContext, useEffect } from 'react';
import axios, { AxiosResponse } from 'axios';

import { Store } from '../../Context/Store';
import { Theme } from '../../Context/ThemeContext';

// actions
import { FETCH_SUMMARY, FETCH_BY_COUNTRY, FETCH_COUNTRIES, SWITCH_THEME_NIGHT } from "../../Actions/MainActions";

// const
import { fetchSummary, fetchByCountryTotal, fetchCountries as fetchCountriesURL } from "../../Assets/Conts/fetchURL";
import { initialArrayLength } from "../../Assets/Conts/numbers";

// interfaces and types
import { FormElement } from "../../Types/Types";

// components
import Headline from "../../components/Headline";
import TextPanel from "../../components/TextPanel";
import ChooseCountryPanel from "../../components/ChooseCountryPanel";
import Table from "../../components/Table";

// styles
import { Container, Grid } from "./App.style";

// images
import CovidBg from "../../Assets/Images/Backgrounds/covid-bg.jpg";

function App(): JSX.Element {
  const { state, dispatch } = useContext(Store);
  const { state: themeState, dispatch: themeDispatch } = useContext(Theme);
  const [currentValue, setCurrentValue] = useState<string>('');

  console.log('THEME: ', themeState);
  const handleChange = (event: FormElement): void => {
    setCurrentValue(event.currentTarget.value);
  };

  const handleChangeTheme = (): void => {
    SWITCH_THEME_NIGHT('night', themeDispatch);
  };

  const fetchSummaryByCountry = async () => {
    const data: AxiosResponse = await axios.get(`${fetchByCountryTotal}/${currentValue}/status/confirmed`);
    FETCH_BY_COUNTRY(data, dispatch);
  };

  const fetchSummaryData = async () => {
    const data: AxiosResponse = await axios.get(fetchSummary);
    FETCH_SUMMARY(data.data, dispatch);
  };

  const fetchCountries = async () => {
    const data: AxiosResponse = await axios.get(fetchCountriesURL);
    FETCH_COUNTRIES(data, dispatch);
  };

  useEffect(() => {
    state.allData.length === initialArrayLength && fetchSummaryData();
    state.countries.length === initialArrayLength && fetchCountries();
  }, );

  return (
    <Container maxWidth="xl" disableGutters>
      <Grid container justify="center">
        <Grid item xs={12} section="true" image={CovidBg}>
          <Grid container justify="center">
            {state.allData.length !== initialArrayLength && (
              <>
                <Grid item xs={12}>
                  <Headline totalCount={state.allData.Global.TotalConfirmed} />
                </Grid>
                <Grid item xs={12} sm={2}>
                  <TextPanel title="Total death" count={state.allData.Global.TotalDeaths} status="death"/>
                </Grid>
                <Grid item xs={12} sm={2}>
                  <TextPanel title="Total overcame" count={state.allData.Global.TotalRecovered} status="overcame"/>
                </Grid>
              </>
            )}
          </Grid>
        </Grid>
        <Grid item xs={12} section="true">
          <Grid container justify="center">
            <Grid item xs={12}>
              <button type="button" onClick={handleChangeTheme}>Theme</button>
            </Grid>
            {state.countries.length !== initialArrayLength && (
              <Grid item xs={12}>
                <ChooseCountryPanel countries={state.countries.data} handleChange={handleChange} currentValue={currentValue} submitFunc={fetchSummaryByCountry}/>
              </Grid>
            )}
            {state.countryData.length !== initialArrayLength && (
              <Grid item xs={12}>
                <Table data={state.countryData.data}/>
              </Grid>
            )}
          </Grid>
        </Grid>
      </Grid>
    </Container>
  );
}

export default App;
