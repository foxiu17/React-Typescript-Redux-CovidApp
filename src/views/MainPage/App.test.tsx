import React from 'react';
import { shallow, mount, ReactWrapper, ShallowWrapper } from "enzyme";

import { StoreProvider } from "../../Context/Store";

import App from './App';
import Headline from '../../components/Headline';
import TextPanel from '../../components/TextPanel';

import { Container, Grid } from "./App.style";

describe('Rendering App component - ', () => {
  test('Initial App render', () => {
    const component: ReactWrapper = mount(<StoreProvider><App /></StoreProvider>);
    const container: ReactWrapper = component.find(Container);
    expect(component.exists()).toBe(true);
    expect(container.exists()).toBe(true);
    expect(component).toMatchSnapshot();
  });
});
