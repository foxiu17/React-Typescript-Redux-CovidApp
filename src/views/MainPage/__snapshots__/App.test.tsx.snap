// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering App component -  Initial App render 1`] = `ReactWrapper {}`;
